### DISCUSSION TITLE
I‘m trying to run the hpipm example of simulink, error with qp_data.c

### DISCUSSION THREAD

**Discussion by ufxzp** [giaf/hpipm#167] [Q&A]
I followed the README of hpipm to run the simulink example. When I run `make_sfun.m`, error reported in matlab command window as followed. The qp_data.c(which provided by the example) contains some empty matrix parameters, how can it be compiled? Thanks!
------------------------------------------------------------------
Building with 'Microsoft Visual C++ 2019 (C)'.
cl /c /Zp8 /GR /W3 /EHs /nologo /MD /O2 /Oy- /DNDEBUG /DMATLAB_DEFAULT_RELEASE=R2017b  /DUSE_MEX_CMD   /D_CRT_SECURE_NO_DEPRECATE /D_SCL_SECURE_NO_DEPRECATE /D_SECURE_SCL=0   /DMATLAB_MEX_FILE -I"C:\Users\s9fptz\source\repos\acados\external\hpipm\include" -I"C:\Users\s9fptz\source\repos\acados\external\blasfeo\include"  -I"C:\Program Files\MATLAB\R2020b\extern\include" -I"C:\Program Files\MATLAB\R2020b\simulink\include" "C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c" /FoC:\Users\s9fptz\AppData\Local\Temp\mex_230145169331919_13828\qp_data.obj
Error using mex
qp_data.c
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(147): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(171): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(195): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(223): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(224): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(225): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(226): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(227): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(228): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(229): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(230): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(231): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(232): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(233): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(234): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(235): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(236): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(237): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(238): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(239): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(240): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(241): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(242): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(243): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(247): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(248): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(249): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(250): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(251): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(252): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(253): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(254): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(255): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(256): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(257): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(258): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(259): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(260): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(261): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(262): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(263): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(264): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(265): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(266): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(267): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(271): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(272): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(273): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(274): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(275): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(276): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(277): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(278): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(279): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(280): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(281): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(282): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(283): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(284): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(285): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(286): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(287): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(288): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(289): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(290): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(291): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(295): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(296): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(297): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(298): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(299): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(300): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(301): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(302): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(303): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(304): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(305): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(306): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(307): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(308): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(309): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(310): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(311): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(312): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(313): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(314): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(315): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(319): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(320): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(321): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(322): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(323): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(324): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(325): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(326): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(327): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(328): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(329): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(330): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(331): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(332): error C2059: syntax error: '}'
C:\Users\s9fptz\source\repos\acados\external\hpipm\examples\simulink\qp_data.c(332): fatal error C1003: error count exceeds 100; stopping compilation

Error in make_sfun (line 17)
eval( [ 'mex -v -output  hpipm_solver_sfunction ', ...

**Comment by giaf** [comment]
Hi, this seems something specific to your compiler.
I would suggest you to modify the file `qp_data.c` by hand e.g. at line 147 until you manage to make the compiler happy.
You can try to replace
```static double S20[] = {};```
e.g. with
```static double S20[] = { , } ;```
or
```static double S20[];```
or
```static double *S20;```
or
```static double *S20 = NULL;```
or whatever.
Once you have it working please post here the solution, so I can see if it is acceptable also for other compilers and in case to make it as the default one in the routine generating that code.

**Reply by ufxzp** [reply]
Hi, thanks for your reply!
I tried to replace  
`static double S20[] = { };` 
with many other expressions, but my compiler still reports errors.
I have to replace it with 
`static double S20[] = {0};`
as a temporary solution, which may not be correct... 
I‘m not sure what will go wrong when I replace all the empty parts in [qp_data.c] with '0'?

Anyway, now 'make_sfun.m' runs successfully, the next step is to run `load_parameters.m`.
But a new error reported:

> Error using mex
> ocp_qp_dim_load.c
> C:\Users\s9fptz\source\repos\acados\external\hpipm\interfaces\matlab_octave\ocp_qp_dim_load.c(50): fatal error C1083: Cannot open include file: 'QP_DATA_H': No such file or directory
> 
> 
> 
> Error in compile_mex_one_from_script (line 109)
> 		mex(mex_flags, ['CFLAGS=\$CFLAGS -std=c99', varargin{2}], hpipm_include, blasfeo_include, hpipm_lib, blasfeo_lib, data_include, '-lhpipm', '-lblasfeo', mex_file);
> 
> Error in hpipm_ocp_qp_dim (line 48)
> 				compile_mex_one_from_script('ocp_qp_dim_load.c', [' -DQP_DATA_H=', file]);
> 
> Error in load_parameters (line 17)
> dim = hpipm_ocp_qp_dim('qp_data.c');

Seems like that in 'ocp_qp_dim_load.c', `#include STR(QP_DATA_H)` where the 'QP_DATA_H' can not be found.
Could you please help me with the problem? Thank you~

**Comment by giaf** [comment]
The solution you found `static double S20[] = {0};` is fine, whatever value is in there would make the job as it is not accessed anyway.

Regarding the next issue looks like `QP_DATA_H` is not resolved, I don't know why.
You can try to hard-code the data file name in here https://github.com/giaf/hpipm/blob/master/interfaces/matlab_octave/ocp_qp_dim_load.c#L50
